File: scripts/release/ca_registry_smoke.mjs
// SPDX-FileCopyrightText: 2026 The Attestplane Authors
// SPDX-License-Identifier: Apache-2.0
/**
 * Clean-install smoke for a *published* @attestplane/attestplane release.
 *
 * Evidence for the Controlled Availability admission criteria (#3 clean-install
 * smoke and the verifier portion of #4), run against the package installed from
 * npm — not the local source tree. Run from a throwaway project that has
 * `npm install @attestplane/attestplane@<version>`.
 *
 * It imports only the public package entrypoint. Modes:
 *   * `smoke [EXPECTED_VERSION]` (default): assert VERSION matches the expected
 *     version (catches a stale-literal version drift) and is not the
 *     manifest-missing sentinel; build a bundle and verify -> ok; tamper and
 *     verify -> rejected (claim-safety invariant).
 *   * `emit <PATH>`: write a valid bundle JSON for the cross-SDK roundtrip.
 *   * `verify <PATH>`: verify a bundle JSON produced by the other SDK -> ok.
 *
 * Usage:
 *   node ca_registry_smoke.mjs [smoke [EXPECTED_VERSION] | emit PATH | verify PATH]
 */

import { readFileSync, writeFileSync } from 'node:fs';

import {
  AttestSubstrate,
  ProofBundleBuilder,
  VERSION,
  makeEventDraft,
  verifyProofBundle,
} from '@attestplane/attestplane';

function fail(msg) {
  console.error(`::error::${msg}`);
  process.exit(1);
}

function buildValidBundle() {
  const sub = new AttestSubstrate();
  sub.append(makeEventDraft({ event_type: 'eval_event', actor: 'agent' }), {
    now: new Date('2026-01-01T00:00:00.000Z'),
    event_id: '00000000-0000-7000-8000-000000000001',
  });
  const builder = new ProofBundleBuilder({ chain_id: 'ca-smoke', producer_runtime: 'ca-smoke' });
  builder.extend(sub.snapshot());
  return builder.build();
}

const mode = process.argv[2] ?? 'smoke';

if (mode === 'emit') {
  const path = process.argv[3];
  if (!path) fail('emit requires an output path');
  writeFileSync(path, JSON.stringify(buildValidBundle()), 'utf8');
  console.log(`CA emit OK: wrote bundle to ${path}`);
} else if (mode === 'verify') {
  const path = process.argv[3];
  if (!path) fail('verify requires an input path');
  const bundle = JSON.parse(readFileSync(path, 'utf8'));
  if (!verifyProofBundle(bundle).ok) fail(`cross-SDK bundle failed verification: ${path}`);
  console.log(`CA cross-verify OK: @attestplane/attestplane ${VERSION} verified ${path}`);
} else {
  // Default: self smoke. Accept "smoke [VERSION]" or a bare "[VERSION]".
  const expected = mode === 'smoke' ? process.argv[3] : mode;

  if (VERSION === '0.0.0+unknown') {
    fail('VERSION is the manifest-missing sentinel');
  }
  if (expected !== undefined && VERSION !== expected) {
    fail(`VERSION '${VERSION}' != expected '${expected}' (version drift)`);
  }

  const bundle = buildValidBundle();
  if (!verifyProofBundle(bundle).ok) {
    fail('valid bundle failed verification');
  }

  const tampered = JSON.parse(JSON.stringify(bundle));
  tampered.events[0].event.actor = 'agent://tampered';
  if (verifyProofBundle(tampered).ok) {
    fail('tampered bundle verified ok (claim-safety broken)');
  }

  console.log(`CA registry smoke OK: @attestplane/attestplane ${VERSION} (valid->ok, tampered->rejected)`);
}

File: scripts/release/ca_registry_smoke.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2026 The Attestplane Authors
# SPDX-License-Identifier: Apache-2.0
"""Clean-install smoke for a *published* attestplane release.

Evidence for the Controlled Availability admission criteria (#3 clean-install
smoke and the verifier portion of #4), run against the package installed from
PyPI — not a source checkout.

It imports only the public API. Modes:
  * ``smoke [EXPECTED_VERSION]`` (default): assert the import resolves to a
    site-packages install (not source) and, when a version is given, that
    ``__version__`` matches it (catches a stale-literal version drift); build a
    bundle and verify it -> ok; tamper it and verify -> rejected (claim-safety
    invariant: a mutated bundle must never verify ok).
  * ``emit <PATH>``: write a valid bundle JSON for the cross-SDK roundtrip
    (the *other* SDK, installed from its registry, verifies it).
  * ``verify <PATH>``: verify a bundle JSON produced by the other SDK -> ok.

This is intentionally a focused public-surface smoke, not the full source
conformance suite (whose vectors are not shipped in the wheel). The byte-frozen
conformance suite runs against source in CI, and the published wheel is the
byte-identical reproducible build of that source.

Usage:
    python ca_registry_smoke.py [smoke [EXPECTED_VERSION] | emit PATH | verify PATH]
"""

from __future__ import annotations

import copy
import json
import sys
from datetime import UTC, datetime
from pathlib import Path

import attestplane


def _fail(msg: str) -> int:
    print(f"::error::{msg}", file=sys.stderr)
    return 1


def _build_valid_bundle() -> dict:
    from attestplane import AttestSubstrate, EventDraft, ProofBundleBuilder

    sub = AttestSubstrate()
    base = datetime(2026, 1, 1, tzinfo=UTC)
    chain = [sub.append(EventDraft(event_type="eval_event", actor="agent", payload={"i": 0}), now=base)]
    builder = ProofBundleBuilder(chain_id="ca-smoke", producer_runtime="ca-smoke")
    builder.extend(chain)
    return builder.build(now=base)


def main(argv: list[str]) -> int:
    mode = argv[1] if len(argv) > 1 else "smoke"
    from attestplane import verify_proof_bundle

    if mode == "emit":
        if len(argv) < 3:
            return _fail("emit requires an output path")
        Path(argv[2]).write_text(json.dumps(_build_valid_bundle()), encoding="utf-8")
        print(f"CA emit OK: wrote bundle to {argv[2]}")
        return 0

    if mode == "verify":
        if len(argv) < 3:
            return _fail("verify requires an input path")
        bundle = json.loads(Path(argv[2]).read_text(encoding="utf-8"))
        result = verify_proof_bundle(bundle)
        if not result.ok:
            return _fail(f"cross-SDK bundle failed verification: {result}")
        print(f"CA cross-verify OK: attestplane {attestplane.__version__} verified {argv[2]}")
        return 0

    # Default: self smoke. Accept "smoke [VERSION]" or a bare "[VERSION]".
    if mode == "smoke":
        expected = argv[2] if len(argv) > 2 else None
    else:
        expected = mode  # bare version argument

    # 1. Identity guard: must be the installed wheel, never a source tree.
    location = attestplane.__file__
    if "site-packages" not in location:
        return _fail(f"attestplane resolved to non-install path: {location}")
    if expected is not None and attestplane.__version__ != expected:
        return _fail(
            f"attestplane.__version__ {attestplane.__version__!r} "
            f"!= expected {expected!r} (version drift)"
        )
    if attestplane.__version__ == "0.0.0+unknown":
        return _fail("attestplane.__version__ is the dist-info-missing sentinel")

    # 2. Build a valid bundle with the public API and verify -> ok.
    bundle = _build_valid_bundle()
    if not verify_proof_bundle(bundle).ok:
        return _fail("valid bundle failed verification")

    # 3. Tamper -> reject (claim-safety: a mutated bundle must not verify ok).
    tampered = copy.deepcopy(bundle)
    tampered["events"][0]["event"]["payload"]["i"] = 999
    if verify_proof_bundle(tampered).ok:
        return _fail("tampered bundle verified ok (claim-safety broken)")

    print(
        f"CA registry smoke OK: attestplane {attestplane.__version__} @ {location} "
        f"(valid->ok, tampered->rejected)"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
